File: Cloud Storage Samples/ShareFile/java/sharefile.java
/*
 * Cloud Storage 2022 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import cloudstorage.*;

public class sharefile extends ConsoleDemo {
    Sharefile sharefile1 = new Sharefile();
    Oauth oauth1 = new Oauth();
    String currentFolderId = "";
    String subdomain = "";

    public sharefile() {

        String buffer = "";
        try {
			System.out.println("\n*****************************************************************");
			System.out.println("\n* This demo shows how to use the ShareFile component to list,   *");
			System.out.println("\n* upload, download, delete, and share documents from ShareFile. *");
			System.out.println("\n* Default prompt values are in [ ] where applicable.            *");
			System.out.println("\n*****************************************************************\n");
			
            /* NOTE: ShareFile requires a registered redirect URI which is https://
            *       and explicitly disallows localhost. A valid public domain with SSL enabled must
            *       be specific when creating your ClientId and ClientSecret values
            *       
            *       Visit https://api.sharefile.com/apikeys to create your credentials
            *       
            * IMPORTANT: 
            *      Because of the ShareFile restrictions detailed above this demo
            *      make use of an existing public URL (oauth.nsoftware.com)
            *      This should NOT be used in production. It is only used here to
            *      facilitate running of this demo. In a real application you should
            *      register your own Redirect URI.
            *      
            *      For any questions please contact support@nsoftware.com
            */

            subdomain = prompt("Account Subdomain",":","");

            oauth1.setClientProfile(0); // Application
            oauth1.setClientId(prompt("OAuth Client ID",":","Hb7kR4d6mBvQxnOe9Eg6V8CMIVjQ67vD"));
            oauth1.setClientSecret(prompt("OAuth Client Secret",":","jHAZH6f1Ac5O9krW9az3fLYFEZW4KoF1sr1hqDoVgrvv1zwc"));
            oauth1.setServerAuthURL("https://secure.sharefile.com/oauth/authorize");
            oauth1.setServerTokenURL("https://" + subdomain + ".sharefile.com/oauth/token");

            oauth1.setWebServerPort(7878); // Specify a particular port, as it will be used in the redirect.
            oauth1.startWebServer(); // In most cases, the webserver is automatically started when getAuthorization is called.  In this case, due to the use of oauth.nsoftware.com, it is started manually.
            oauth1.setReturnURL("https://oauth.nsoftware.com/oauthdemo"); // This URL should NOT be used in a production application.  It is only for testing purposes.
            oauth1.addParam("state", "http://localhost:" + oauth1.getWebServerPort()); // The oauth.nsoftware.com server uses this value to redirect to the local server embedded within the OAuth class.
            
			//oauth1.setRefreshToken(""); // Set to value from previous runs to expedite the login process.
            sharefile1.setAuthorization(oauth1.getAuthorization());
            //System.out.println(oauth1.getRefreshToken()); // RefreshToken may be used to automatically obtain a new token without user interaction.

            System.out.println("Home Directory Listings: ");
            listItems("home");
            printMenu();
            while (!buffer.equals("q")) {
                System.out.print("Enter command: ");
                buffer = input();
                switch (buffer){
                  case "ls":
                    listItems(currentFolderId);
                    break;
                  case "cd":
                    System.out.print("Number corresponding to folder you wish to view: ");
                    int folderSelection = Integer.valueOf(input());
                    if(folderSelection >= 0){
                      listItems(sharefile1.getItems().get(folderSelection).getId());
                    } else {
                      sharefile1.getItemInfo(currentFolderId);
                      listItems(sharefile1.getItems().get(0).getParentId());
                    }                 
                    break;
                  case "get":
                    System.out.print("Number corresponding to item you wish to download: ");
                    ShareFileItem downloadSelection = sharefile1.getItems().get(Integer.valueOf(input()));
                    System.out.print("Directory in which to download: ");

                    sharefile1.setLocalFile(input() + "\\" + downloadSelection.getName());
                    sharefile1.downloadFile(downloadSelection.getId()); 

                    System.out.println("\nDownload complete.\n");
                    break;
                  case "put":
                    System.out.print("Path of file that will be uploaded: ");
                    sharefile1.setLocalFile(input());

                    File mFile = new File(sharefile1.getLocalFile());
                    sharefile1.uploadFile(mFile.getName(), currentFolderId);

                    listItems(currentFolderId);
                    break;
                  case "del":
                    System.out.print("Number corresponding to item you wish to delete: ");
                    ShareFileItem deleteSelection = sharefile1.getItems().get(Integer.valueOf(input()));

                    sharefile1.deleteItem(deleteSelection.getId());

                    listItems(currentFolderId);
                    break;
                  case "mkdir":
                    System.out.print("Name of new folder: ");
                    sharefile1.createFolder(input(), currentFolderId);
                    listItems(currentFolderId);
                    break;
                  case "share":
                    System.out.print("Number corresponding to item you wish to share: ");
                    ShareFileItem shareSelection = sharefile1.getItems().get(Integer.valueOf(input()));
                    System.out.println("Share Link: " + sharefile1.createLink(shareSelection.getId()));
                    break;
                  case "request":
                    System.out.println("Request Link: " + sharefile1.createRequestLink(currentFolderId));
                    break;
                  case "q":
                    System.exit(0);
                    break;
                  default:
                    printMenu();
                } 
            }
        } catch (CloudStorageException e) {
            System.out.println(e.getMessage());
            System.exit(e.getCode());
            return;
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

    }

    private void printMenu() {
        System.out.println("\n?\t-\tHelp\n" + "ls\t-\tList Items\n"
                + "cd\t-\tChange Folders\n" + "get\t-\tDownload Item\n" 
                + "put\t-\tUpload Item\n" + "del\t-\tDelete Item\n" 
                + "mkdir\t-\tCreate Folder\n" + "share\t-\tShare Items\n" 
                + "request\t-\tRequest Items\n" + "q\t-\tQuit\n");
    }

    private void listItems(String FolderId) {
        try {
            sharefile1.setAccountSubdomain(subdomain);
            sharefile1.getItemInfo(FolderId);
            currentFolderId = sharefile1.getItems().get(0).getId();

            System.out.print("\n");
            System.out.format("%1$-3s%2$-20s%3$-30s%4$-40s\n","#","Title","Type","Creation Date");
            System.out.println("--------------------------------------------------------------------------------");
            
            sharefile1.listItems(currentFolderId);
            System.out.format("%1$-3s%2$-20.18s%3$-30s%4$-40s\n", "-1", "..", "Folder", "");
            for (int i = 0; i < sharefile1.getItems().size(); i++) {
                ShareFileItem currentItem = sharefile1.getItems().get(i);
                System.out.format("%1$-3s%2$-20.18s%3$-30s%4$-40s\n", String.valueOf(i), currentItem.getName(),
                    ResolveItemType(currentItem.getType()), currentItem.getCreationDate());
       }

            System.out.print("\n");

        } catch (CloudStorageException e) {
            System.out.println(e.getMessage());
            System.exit(e.getCode());
            return;
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }

    private String ResolveItemType(int type) {
        switch (type) {
            case ShareFileItem.sfitFile:	return "File";
            case ShareFileItem.sfitFolder: return "Folder";
            default: return "Unknown";
        }
    }

    public static void main(String[] args) {
        try {
            new sharefile();
        } catch (Exception ex) {
            System.out.println("Exception: " + ex.getMessage());
        }
    }

}
class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudStorageException) {
      System.out.print(" (" + ((CloudStorageException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

File: Cloud Storage Samples/ShareFile/c++/sharefile.cpp
/*
 * Cloud Storage 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/cloudstorage.h"
#include <string>
#define LINE_LEN 255

OAuth oauth1;
ShareFile sharefile1;
enum { IT_FILE, IT_FOLDER }; // Enums defining the item type.
char currentFolderId[LINE_LEN + 1]; // The ID of the folder that the user is currently in.

/*
* Method to print the error codes
*/
void print_error(int ret_code) {
	printf("\nError: %d", ret_code);
	if (sharefile1.GetLastError())
	{
		printf(" \"%s\"\n", sharefile1.GetLastError());
	}
	if (oauth1.GetLastError()) {
		printf(" \"%s\"\n", oauth1.GetLastError());
	}
	printf("Press any key to continue.");
	fflush(stdin);
	getchar();
	exit(ret_code);
}

/*
* Method to print the menu options
*/
void printOptions()
{
	printf( "Type ? to view this menu, or type one of the numbers below to perform the corresponding operation:\n"
	        "1.  List Items\n"
	        "2.  Change Folders\n"
	        "3.  Download Items\n"
	        "4.  Upload Items\n"
	        "5.  Delete Items\n"
	        "6.  Create Folder\n"
	        "7.  Create Request Link\n"
	        "8.  Create Share Link\n"
	        "9.  Quit\n");
}

/*
* Method that converts the item type enum to a readable string.
*/
char* resolveItemType(int type) {
  switch (type)
  {
    case IT_FILE: return "File";
    case IT_FOLDER: return "Folder";
    default: return "Unknown";
  }
}

/*
* Method that lists the items in the passed folder.  Updates currentFolderId.
*/
void displayItems(const char* folderId){

	sharefile1.GetItemInfo(folderId);
	strcpy(currentFolderId, sharefile1.GetItemId(0));
	int ret_code = sharefile1.ListItems(currentFolderId);
	if(ret_code)
		print_error(ret_code);

	printf("%-3.1s%-40.38s%-10.8s%-25.23s\n\n","#","Title","Type","Creation Date");
	
	for(int i=0;i<sharefile1.GetItemCount();i++)
	{
		printf("%-3.1d%-40.38s%-10.8s%-25.23s\n",i, sharefile1.GetItemName(i),resolveItemType(sharefile1.GetItemType(i)),sharefile1.GetItemCreationDate(i));
	}
	return;
}

/*
* Main Method
*/
int main(int argc, char **argv)
{
	char subdomain[LINE_LEN + 1] = "";
	char buffer[LINE_LEN + 1] = "";
	int ret_code = 0;

  printf("\n*****************************************************************");
  printf("\n* This demo shows how to use the ShareFile component to list,   *");
  printf("\n* upload, download, delete, and share documents from ShareFile. *");
  printf("\n* Default prompt values are in [ ] where applicable.            *");
  printf("\n*****************************************************************\n");

  printf("Account  Subdomain: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer) - 1] = '\0';
  if (strlen(buffer) != 0) {
	  strcpy(subdomain, buffer);
  } else {
	  printf("Must set a subdomain (press any key to quit): ");
	  fgets(buffer, LINE_LEN, stdin);
	  return -1;
  }

  printf("Client ID [Hb7kR4d6mBvQxnOe9Eg6V8CMIVjQ67vD]: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  if (strlen(buffer) == 0) { 
    oauth1.SetClientId("Hb7kR4d6mBvQxnOe9Eg6V8CMIVjQ67vD");
  } else { 
    oauth1.SetClientId(buffer);
  }
  
  printf("Client Secret [jHAZH6f1Ac5O9krW9az3fLYFEZW4KoF1sr1hqDoVgrvv1zwc]: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  if (strlen(buffer) == 0) {
    oauth1.SetClientSecret("jHAZH6f1Ac5O9krW9az3fLYFEZW4KoF1sr1hqDoVgrvv1zwc");
  } else {
    oauth1.SetClientSecret(buffer);
  }
  
  oauth1.SetClientProfile(0); // Application
  oauth1.SetServerAuthURL("https://secure.sharefile.com/oauth/authorize");
  
	//Build the ServerTokenURL string with the specified subdomain
	strcpy(buffer, "https://");
	strcat(buffer, subdomain);
    strcat(buffer, ".sharefile.com/oauth/token");
	buffer[strlen(buffer)] = '\0';
    oauth1.SetServerTokenURL(buffer);
	
	//Web Server (port in param must equal port in oauth1.SetWebServerPort())
	char* param = "http://localhost:7878";
	oauth1.SetWebServerPort(7878); // Specify a particular port, as it will be used in the redirect.
	oauth1.StartWebServer(); // In most cases, the webserver is automatically started when getAuthorization is called.  In this case, due to the use of oauth.nsoftware.com, it is started manually.
	oauth1.SetReturnURL("https://oauth.nsoftware.com/oauthdemo"); // This URL should NOT be used in a production application.  It is only for testing purposes.
	oauth1.AddParam("state", param); // The oauth.nsoftware.com server uses this value to redirect to the local server embedded within the OAuth class.
	//oauth1.SetRefreshToken(""); // Set to value from previous runs to expedite the login process.
    sharefile1.SetAuthorization(oauth1.GetAuthorization());
	//printf("Refresh Token: %s", oauth1.GetRefreshToken()); // RefreshToken may be used to automatically obtain a new token without user interaction.

	ret_code = oauth1.GetLastErrorCode();

	if (ret_code) {
		print_error(ret_code);
	}
 
	sharefile1.SetAccountSubdomain(subdomain);
    printf("\n");
	printf("Home Directory Listings: \n");
	displayItems("home");

	printOptions();

	while (1)
	{
		printf( "\n> " );
		scanf("%80s",buffer);

		if (!strcmp(buffer, "?") ) // List options.
		{
			printOptions();
		}

		else if (!strcmp(buffer, "1") ) // List items.
		{
			displayItems("home");
		}
		
		else if (!strcmp(buffer, "2")) // Change folders.
		{
			printf("\nNumber corresponding to folder you wish to view: ");
			scanf("%80s", buffer);
			int index = atoi(buffer);
			if (index < 0) {
				sharefile1.GetItemInfo(currentFolderId);
				strcpy(buffer, sharefile1.GetItemParentId(0));
				displayItems(buffer);
			}
			else {
				if (sharefile1.GetItemType(index) == IT_FOLDER) {
					strcpy(buffer, sharefile1.GetItemId(index));
					displayItems(buffer);
				}
				else {
					printf("\nPlease choose a valid directory.\n");
				}
			}
		}
		
		else if (!strcmp(buffer, "3") ) // Download item.
		{
			printf("\nNumber corresponding to item you wish to download: ");
			scanf("%80s",buffer);
			int index = atoi(buffer);
			
			printf("Directory in which to download: ");
			scanf("%80s",buffer);
			strcat(buffer,"\\");
			strcat(buffer,sharefile1.GetItemName(index));
			sharefile1.SetLocalFile(buffer);

			ret_code = sharefile1.DownloadFile(sharefile1.GetItemId(index));
			if(ret_code)
				print_error(ret_code);

			printf("\nDownload complete.\n");
		}

		else if (!strcmp(buffer, "4") ) // Upload file.
		{
			printf("\nPath of file that will be uploaded: ");
			scanf("%80s",buffer);

			sharefile1.SetLocalFile(buffer);
			
      sharefile1.UploadFile(sharefile1.GetLocalFile(), currentFolderId);
			ret_code = sharefile1.GetLastErrorCode();

			if(ret_code)
				print_error(ret_code);

			displayItems(currentFolderId);
		}

		else if (!strcmp(buffer, "5") ) // Delete item.
		{
			printf("\nNumber corresponding to item you wish to delete: ");
			scanf("%80s",buffer);
			int index = atoi(buffer);
			ret_code = sharefile1.DeleteItem(sharefile1.GetItemId(index));
			if(ret_code)
				print_error(ret_code);

			displayItems(currentFolderId);
		}

		else if (!strcmp(buffer, "6")) // Create folder.
		{
			printf("\nName of new folder: ");
			scanf("%80s", buffer);
			sharefile1.CreateFolder(buffer, currentFolderId);
			displayItems(currentFolderId);
		}

		else if (!strcmp(buffer, "7")) // Create request link.
		{
			printf("Request Link: %s\n", sharefile1.CreateRequestLink(currentFolderId));

		}

		else if (!strcmp(buffer, "8")) // Create share link.
		{
		printf("\nNumber corresponding to item you wish to share: ");
		scanf("%80s", buffer);
		int index = atoi(buffer);
		printf("Share Link: %s\n", sharefile1.CreateLink(sharefile1.GetItemId(index)));
		}

		else if (!strcmp(buffer, "9") )
		{
			exit(0);
		}

		else if (!strcmp(buffer, "") )
		{
			// Do nothing.
		}
		else
		{
			printf("\nPlease select a number from 1-9 or ? for possible commands.\n");
		}
	}
	return ret_code;
}

